language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -220,47 +242,86 @@ typeOf (TyInst x body ty) = do
+    tyEnvAssign (extractUnique n) (NormalizedType $ TyFix () (void n) (void ty))
     typeCheckStep
-    if tyReduce fixed == nBodyTy
-        then pure $ NormalizedType (TyFix () (void n) (void ty)) -- type annotations on terms must be normalized
-        else throwError (TypeMismatch x (void body) fixed nBodyTy)
+    red <- tyReduce (void ty) <* tyEnvDelete (extractUnique n)
+    if red == nBodyTy
+        then pure $ NormalizedType (TyFix () (void n) (void ty))
+        else throwError (TypeMismatch x (void body) (getNormalizedType red) nBodyTy)
 
 extractUnique :: TyNameWithKind a -> Unique
 extractUnique = nameUnique . unTyName . unTyNameWithKind
 
--- TODO: make type substitutions occur in a state monad + benchmark
--- | Substitute the given type into another type. The input types must be normalized - this prevents us creating redundant work. The
--- output type is not normalized, since we may introduce new redexes.
-tySubstitute :: Unique -- ^ Unique associated with type variable
-             -> NormalizedType TyNameWithKind a -- ^ Type we are binding to free variable
-             -> NormalizedType TyNameWithKind a -- ^ Type we are substituting in
-             -> Type TyNameWithKind a
-tySubstitute u (NormalizedType ty) = cata a . getNormalizedType where
-    a (TyVarF _ (TyNameWithKind (TyName (Name _ _ u')))) | u == u' = ty
-    a x                                                  = embed x
-
--- also this should involve contexts
+tyEnvDelete :: MonadState (TypeSt, Natural) m
+            => Unique
+            -> m ()
+tyEnvDelete (Unique i) = modify (first (IM.delete i))
+
+tyEnvAssign :: MonadState (TypeSt, Natural) m
+            => Unique
+            -> NormalizedType TyNameWithKind ()
+            -> m ()
+tyEnvAssign (Unique i) ty = modify (first (IM.insert i ty))
+
+isTyLam :: Type TyNameWithKind () -> Bool
+isTyLam TyLam{} = True
+isTyLam _       = False
+
+maybeRed :: Bool -> Type TyNameWithKind () -> TypeCheckM a (NormalizedType TyNameWithKind ())
+maybeRed True  = tyReduce
+maybeRed False = pure . NormalizedType
+
+-- This function is necessary because type reductions do *not* traverse 'TyFix'
+-- but context rewrites must traverse *all* constructors.
+rewriteCtx :: Type TyNameWithKind () -> TypeCheckM a (Type TyNameWithKind ())
+rewriteCtx (TyApp x ty ty')     = TyApp x <$> rewriteCtx ty <*> rewriteCtx ty'
+rewriteCtx (TyFun x ty ty')     = TyFun x <$> rewriteCtx ty <*> rewriteCtx ty'
+rewriteCtx (TyFix x tn ty')     = TyFix x tn <$> rewriteCtx ty'
+rewriteCtx (TyLam x tn k ty)    = TyLam x tn k <$> rewriteCtx ty
+rewriteCtx (TyForall x tn k ty) = TyForall x tn k <$> rewriteCtx ty
+rewriteCtx ty@TyInt{}           = pure ty
+rewriteCtx ty@TyBuiltin{}       = pure ty
+rewriteCtx ty@(TyVar _ (TyNameWithKind (TyName (Name _ _ u)))) = do
+    (st, _) <- get
+    case IM.lookup (unUnique u) st of
+        Just ty'@(NormalizedType TyVar{}) -> rewriteCtx (getNormalizedType ty')
```

Comment:
Feels like there should be a separate loop that looks up variables in the state until the result is not a variable. Or maybe there should be an invariant that this process happens on variables **insertion**? What the purpose of doing recursive lookups when you can just guarantee that each variable points to an actual type rather than a variable? Or can you not guarantee this? Please comment all this.

Reply:
This should not happen during insertion, mainly because we can have an assignment `a -> b` and an assignment `b -> c` such that the assignment `b -> c` is locally valid but in a smaller scope than the assignment `a -> b`.

Reply:
I understand the words you're saying, but not the actual mechanics. Please justify this design decision in the code itself, so everyone can read a counter-example and realize why this loop on each lookup is required.